In [2]:
# Importing the necessary packages
import pandas as pd
import numpy as np
from imblearn.over_sampling import SMOTE
from sklearn.ensemble import RandomForestClassifier

# To ensure all outputs from a cell are shown
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

# To ensure all columns of a dataframe are shown
pd.set_option("display.max_columns", None)

In [3]:
# Reading in the data
df = pd.read_csv("train.csv")
df

,id,circle_id,loc_og_t2o_mou,std_og_t2o_mou,loc_ic_t2o_mou,last_date_of_month_6,last_date_of_month_7,last_date_of_month_8,arpu_6,arpu_7,arpu_8,onnet_mou_6,onnet_mou_7,onnet_mou_8,offnet_mou_6,offnet_mou_7,offnet_mou_8,roam_ic_mou_6,roam_ic_mou_7,roam_ic_mou_8,roam_og_mou_6,roam_og_mou_7,roam_og_mou_8,loc_og_t2t_mou_6,loc_og_t2t_mou_7,loc_og_t2t_mou_8,loc_og_t2m_mou_6,loc_og_t2m_mou_7,loc_og_t2m_mou_8,loc_og_t2f_mou_6,loc_og_t2f_mou_7,loc_og_t2f_mou_8,loc_og_t2c_mou_6,loc_og_t2c_mou_7,loc_og_t2c_mou_8,loc_og_mou_6,loc_og_mou_7,loc_og_mou_8,std_og_t2t_mou_6,std_og_t2t_mou_7,std_og_t2t_mou_8,std_og_t2m_mou_6,std_og_t2m_mou_7,std_og_t2m_mou_8,std_og_t2f_mou_6,std_og_t2f_mou_7,std_og_t2f_mou_8,std_og_t2c_mou_6,std_og_t2c_mou_7,std_og_t2c_mou_8,std_og_mou_6,std_og_mou_7,std_og_mou_8,isd_og_mou_6,isd_og_mou_7,isd_og_mou_8,spl_og_mou_6,spl_og_mou_7,spl_og_mou_8,og_others_6,og_others_7,og_others_8,total_og_mou_6,total_og_mou_7,total_og_mou_8,loc_ic_t2t_mou_6,loc_ic_t2t_mou_7,loc_ic_t2t_mou_8,loc_ic_t2m_mou_6,loc_ic_t2m_mou_7,loc_ic_t2m_mou_8,loc_ic_t2f_mou_6,loc_ic_t2f_mou_7,loc_ic_t2f_mou_8,loc_ic_mou_6,loc_ic_mou_7,loc_ic_mou_8,std_ic_t2t_mou_6,std_ic_t2t_mou_7,std_ic_t2t_mou_8,std_ic_t2m_mou_6,std_ic_t2m_mou_7,std_ic_t2m_mou_8,std_ic_t2f_mou_6,std_ic_t2f_mou_7,std_ic_t2f_mou_8,std_ic_t2o_mou_6,std_ic_t2o_mou_7,std_ic_t2o_mou_8,std_ic_mou_6,std_ic_mou_7,std_ic_mou_8,total_ic_mou_6,total_ic_mou_7,total_ic_mou_8,spl_ic_mou_6,spl_ic_mou_7,spl_ic_mou_8,isd_ic_mou_6,isd_ic_mou_7,isd_ic_mou_8,ic_others_6,ic_others_7,ic_others_8,total_rech_num_6,total_rech_num_7,total_rech_num_8,total_rech_amt_6,total_rech_amt_7,total_rech_amt_8,max_rech_amt_6,max_rech_amt_7,max_rech_amt_8,date_of_last_rech_6,date_of_last_rech_7,date_of_last_rech_8,last_day_rch_amt_6,last_day_rch_amt_7,last_day_rch_amt_8,date_of_last_rech_data_6,date_of_last_rech_data_7,date_of_last_rech_data_8,total_rech_data_6,total_rech_data_7,total_rech_data_8,max_rech_data_6,max_rech_data_7,max_rech_data_8,count_rech_2g_6,count_rech_2g_7,count_rech_2g_8,count_rech_3g_6,count_rech_3g_7,count_rech_3g_8,av_rech_amt_data_6,av_rech_amt_data_7,av_rech_amt_data_8,vol_2g_mb_6,vol_2g_mb_7,vol_2g_mb_8,vol_3g_mb_6,vol_3g_mb_7,vol_3g_mb_8,arpu_3g_6,arpu_3g_7,arpu_3g_8,arpu_2g_6,arpu_2g_7,arpu_2g_8,night_pck_user_6,night_pck_user_7,night_pck_user_8,monthly_2g_6,monthly_2g_7,monthly_2g_8,sachet_2g_6,sachet_2g_7,sachet_2g_8,monthly_3g_6,monthly_3g_7,monthly_3g_8,sachet_3g_6,sachet_3g_7,sachet_3g_8,fb_user_6,fb_user_7,fb_user_8,aon,aug_vbc_3g,jul_vbc_3g,jun_vbc_3g,churn_probability
0,0,109,0.0,0.0,0.0,6/30/2014,7/31/2014,8/31/2014,31.277,87.009,7.527,48.58,124.38,1.29,32.24,96.68,2.33,0.00,0.0,0.0,0.00,0.0,0.00,2.23,0.00,0.28,5.29,16.04,2.33,0.00,0.00,0.00,0.00,0.00,0.00,7.53,16.04,2.61,46.34,124.38,1.01,18.75,80.61,0.00,0.00,0.0,0.0,0.0,0.0,0.0,65.09,204.99,1.01,0.00,0.0,0.0,8.20,0.63,0.00,0.38,0.0,0.0,81.21,221.68,3.63,2.43,3.68,7.79,0.83,21.08,16.91,0.00,0.00,0.00,3.26,24.76,24.71,0.00,7.61,0.21,7.46,19.96,14.96,0.0,0.0,0.00,0.0,0.0,0.0,7.46,27.58,15.18,11.84,53.04,40.56,0.0,0.0,0.66,0.00,0.0,0.00,1.11,0.69,0.00,3,2,2,77,65,10,65,65,10,6/22/2014,7/10/2014,8/24/2014,65,65,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.00,0.00,0.00,0.00,0.00,0.00,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0,0,0,0,0,0,0,0,0,0,0,0,NaN,NaN,NaN,1958,0.00,0.00,0.00,0
1,1,109,0.0,0.0,0.0,6/30/2014,7/31/2014,8/31/2014,0.000,122.787,42.953,0.00,0.00,0.00,0.00,25.99,30.89,0.00,0.0,0.0,0.00,0.0,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,22.01,29.79,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.0,0.0,0.0,0.0,0.0,0.00,0.00,0.00,0.00,0.0,0.0,0.00,30.73,31.66,0.00,0.0,0.0,0.00,30.73,31.66,1.68,19.09,10.53,1.41,18.68,11.09,0.35,1.66,3.40,3.44,39.44,25.03,0.00,0.00,0.00,0.00,0.00,0.00,0.0,0.0,0.00,0.0,0.0,0.0,0.00,0.00,0.00,3.44,39.44,25.04,0.0,0.0,0.01,0.00,0.0,0.00,0.00,0.00,0.00,3,4,5,0,145,50,0,145,50,6/12/2014,7/10/2014,8/26/2014,0,0,0,NaN,7/8/2014,NaN,NaN,1.0,NaN,NaN,145.0,NaN,NaN,0.0,Na

In [4]:
df.describe()

,id,circle_id,loc_og_t2o_mou,std_og_t2o_mou,loc_ic_t2o_mou,arpu_6,arpu_7,arpu_8,onnet_mou_6,onnet_mou_7,onnet_mou_8,offnet_mou_6,offnet_mou_7,offnet_mou_8,roam_ic_mou_6,roam_ic_mou_7,roam_ic_mou_8,roam_og_mou_6,roam_og_mou_7,roam_og_mou_8,loc_og_t2t_mou_6,loc_og_t2t_mou_7,loc_og_t2t_mou_8,loc_og_t2m_mou_6,loc_og_t2m_mou_7,loc_og_t2m_mou_8,loc_og_t2f_mou_6,loc_og_t2f_mou_7,loc_og_t2f_mou_8,loc_og_t2c_mou_6,loc_og_t2c_mou_7,loc_og_t2c_mou_8,loc_og_mou_6,loc_og_mou_7,loc_og_mou_8,std_og_t2t_mou_6,std_og_t2t_mou_7,std_og_t2t_mou_8,std_og_t2m_mou_6,std_og_t2m_mou_7,std_og_t2m_mou_8,std_og_t2f_mou_6,std_og_t2f_mou_7,std_og_t2f_mou_8,std_og_t2c_mou_6,std_og_t2c_mou_7,std_og_t2c_mou_8,std_og_mou_6,std_og_mou_7,std_og_mou_8,isd_og_mou_6,isd_og_mou_7,isd_og_mou_8,spl_og_mou_6,spl_og_mou_7,spl_og_mou_8,og_others_6,og_others_7,og_others_8,total_og_mou_6,total_og_mou_7,total_og_mou_8,loc_ic_t2t_mou_6,loc_ic_t2t_mou_7,loc_ic_t2t_mou_8,loc_ic_t2m_mou_6,loc_ic_t2m_mou_7,loc_ic_t2m_mou_8,loc_ic_t2f_mou_6,loc_ic_t2f_mou_7,loc_ic_t2f_mou_8,loc_ic_mou_6,loc_ic_mou_7,loc_ic_mou_8,std_ic_t2t_mou_6,std_ic_t2t_mou_7,std_ic_t2t_mou_8,std_ic_t2m_mou_6,std_ic_t2m_mou_7,std_ic_t2m_mou_8,std_ic_t2f_mou_6,std_ic_t2f_mou_7,std_ic_t2f_mou_8,std_ic_t2o_mou_6,std_ic_t2o_mou_7,std_ic_t2o_mou_8,std_ic_mou_6,std_ic_mou_7,std_ic_mou_8,total_ic_mou_6,total_ic_mou_7,total_ic_mou_8,spl_ic_mou_6,spl_ic_mou_7,spl_ic_mou_8,isd_ic_mou_6,isd_ic_mou_7,isd_ic_mou_8,ic_others_6,ic_others_7,ic_others_8,total_rech_num_6,total_rech_num_7,total_rech_num_8,total_rech_amt_6,total_rech_amt_7,total_rech_amt_8,max_rech_amt_6,max_rech_amt_7,max_rech_amt_8,last_day_rch_amt_6,last_day_rch_amt_7,last_day_rch_amt_8,total_rech_data_6,total_rech_data_7,total_rech_data_8,max_rech_data_6,max_rech_data_7,max_rech_data_8,count_rech_2g_6,count_rech_2g_7,count_rech_2g_8,count_rech_3g_6,count_rech_3g_7,count_rech_3g_8,av_rech_amt_data_6,av_rech_amt_data_7,av_rech_amt_data_8,vol_2g_mb_6,vol_2g_mb_7,vol_2g_mb_8,vol_3g_mb_6,vol_3g_mb_7,vol_3g_mb_8,arpu_3g_6,arpu_3g_7,arpu_3g_8,arpu_2g_6,arpu_2g_7,arpu_2g_8,night_pck_user_6,night_pck_user_7,night_pck_user_8,monthly_2g_6,monthly_2g_7,monthly_2g_8,sachet_2g_6,sachet_2g_7,sachet_2g_8,monthly_3g_6,monthly_3g_7,monthly_3g_8,sachet_3g_6,sachet_3g_7,sachet_3g_8,fb_user_6,fb_user_7,fb_user_8,aon,aug_vbc_3g,jul_vbc_3g,jun_vbc_3g,churn_probability
count,69999.000000,69999.0,69297.0,69297.0,69297.0,69999.000000,69999.000000,69999.000000,67231.000000,67312.000000,66296.000000,67231.000000,67312.000000,66296.000000,67231.000000,67312.000000,66296.000000,67231.000000,67312.000000,66296.000000,67231.000000,67312.000000,66296.000000,67231.000000,67312.000000,66296.000000,67231.000000,67312.000000,66296.000000,67231.000000,67312.000000,66296.000000,67231.000000,67312.000000,66296.000000,67231.000000,67312.000000,66296.000000,67231.000000,67312.000000,66296.000000,67231.000000,67312.000000,66296.000000,67231.0,67312.0,66296.0,67231.000000,67312.000000,66296.000000,67231.000000,67312.000000,66296.000000,67231.000000,67312.000000,66296.000000,67231.000000,67312.000000,66296.000000,69999.000000,69999.000000,69999.000000,67231.000000,67312.000000,66296.000000,67231.000000,67312.000000,66296.000000,67231.000000,67312.000000,66296.000000,67231.000000,67312.000000,66296.000000,67231.000000,67312.000000,66296.000000,67231.000000,67312.000000,66296.000000,67231.000000,67312.000000,66296.000000,67231.0,67312.0,66296.0,67231.000000,67312.000000,66296.000000,69999.000000,69999.000000,69999.000000,67231.000000,67312.000000,66296.000000,67231.000000,67312.000000,66296.000000,67231.000000,67312.000000,66296.000000,69999.000000,69999.000000,69999.000000,69999.000000,69999.000000,69999.000000,69999.000000,69999.000000,69999.000000,69999.000000,69999.000000,69999.000000,17568.000000,17865.000000,18417.000000,17568.000000,17865.000000,18417.000000,17568.000000,17865.000000,18417.000000,17568.000000,17865.000000,18417.000000,17568.000000,17865.000000,18417.000000,69999.000000,69999.000000

The columns "circle_id", "loc_og_t2o_mou", "std_og_t2o_mou", "loc_ic_t2o_mou", "std_og_t2c_mou_6", "std_og_t2c_mou_7", "std_og_t2c_mou_8", "std_ic_t2o_mou_6", "std_ic_t2o_mou_7", "std_ic_t2o_mou_8" all have either 0 as the value or the same value for the whole dataset

In [7]:
# Check for columns with missing values
pd.set_option("display.max_rows", None)
df.isna().sum()
pd.set_option("display.max_rows", 10)

id                              0
circle_id                       0
loc_og_t2o_mou                702
std_og_t2o_mou                702
loc_ic_t2o_mou                702
last_date_of_month_6            0
last_date_of_month_7          399
last_date_of_month_8          733
arpu_6                          0
arpu_7                          0
arpu_8                          0
onnet_mou_6                  2768
onnet_mou_7                  2687
onnet_mou_8                  3703
offnet_mou_6                 2768
offnet_mou_7                 2687
offnet_mou_8                 3703
roam_ic_mou_6                2768
roam_ic_mou_7                2687
roam_ic_mou_8                3703
roam_og_mou_6                2768
roam_og_mou_7                2687
roam_og_mou_8                3703
loc_og_t2t_mou_6             2768
loc_og_t2t_mou_7             2687
loc_og_t2t_mou_8             3703
loc_og_t2m_mou_6             2768
loc_og_t2m_mou_7             2687
loc_og_t2m_mou_8             3703
loc_og_t2f_mou

The columns "fb_user_6", "fb_user_7", "fb_user_8", "night_pck_user_6", "night_pck_user_7", "night_pck_user_8", "arpu_3g_6", "arpu_3g_7", "arpu_3g_8", "arpu_2g_6", "arpu_2g_7", "arpu_2g_8","av_rech_amt_data_8", "av_rech_amt_data_7", "av_rech_amt_data_6", "count_rech_3g_8", "count_rech_3g_7", "count_rech_3g_6", "count_rech_2g_8", "count_rech_2g_7", "count_rech_2g_6", "max_rech_data_8", "max_rech_data_7", "max_rech_data_6", "total_rech_data_8", "total_rech_data_7", "total_rech_data_6", "date_of_last_rech_data_8", "date_of_last_rech_data_7", "date_of_last_rech_data_6", "last_date_of_month_6", "last_date_of_month_7", "last_date_of_month_8", all have >51,000 missing values out of 69,999 rows

In [64]:
# List of cols to drop with more than 50% missing values and same values
cols =["circle_id","loc_og_t2o_mou","std_og_t2o_mou","loc_ic_t2o_mou","std_og_t2c_mou_6","std_og_t2c_mou_7","std_og_t2c_mou_8","std_ic_t2o_mou_6","std_ic_t2o_mou_7","std_ic_t2o_mou_8",
      "fb_user_6","fb_user_7","fb_user_8","night_pck_user_6","night_pck_user_7","night_pck_user_8","arpu_3g_6","arpu_3g_7","arpu_3g_8","arpu_2g_6","arpu_2g_7","arpu_2g_8",
      "av_rech_amt_data_8","av_rech_amt_data_7","av_rech_amt_data_6","count_rech_3g_8","count_rech_3g_7","count_rech_3g_6","count_rech_2g_8","count_rech_2g_7","count_rech_2g_6",
      "max_rech_data_8","max_rech_data_7","max_rech_data_6","total_rech_data_8","total_rech_data_7","total_rech_data_6","date_of_last_rech_data_8","date_of_last_rech_data_7","date_of_last_rech_data_6",
      "last_date_of_month_6","last_date_of_month_7","last_date_of_month_8", "id", "date_of_last_rech_6","date_of_last_rech_7","date_of_last_rech_8"]

In [65]:
df= df.drop(cols, axis=1)

In [66]:
# Impute missing values with simple imputer, method is median
from sklearn.impute import SimpleImputer
imp_median = SimpleImputer(missing_values=np.nan, strategy='median')
imp_median.fit(df)
impute_data= pd.DataFrame(imp_median.transform(df))
impute_data.columns= df.columns

SimpleImputer(strategy='median')

In [85]:
# Perform feature selection using RFE
from sklearn.feature_selection import RFE
from sklearn.tree import DecisionTreeClassifier
X2=impute_data.drop(["churn_probability"], axis =1)
y2=impute_data["churn_probability"]
rfe = RFE(estimator=DecisionTreeClassifier(max_depth= 7), n_features_to_select=1)
rfe.fit(X2, y2)

RFE(estimator=DecisionTreeClassifier(max_depth=7), n_features_to_select=1)

In [86]:
# Get the ranks of the column
rank= pd.DataFrame()

rank['column']= X2.columns
rank['rfe_rank']= rfe.ranking_

In [87]:
# Sort the ranks in ascending
rank = rank.sort_values(by= 'rfe_rank')
cols = rank["column"].tolist()
columns_best= cols[0:30] + ["churn_probability"]

In [89]:
# Get the top 30 features
df2=impute_data[columns_best]

In [67]:
from fast_ml.model_development import train_valid_test_split
# Splitting the data into Training, validation and test sets
X_train3, y_train3, X_valid3, y_valid3, X_test3, y_test3 = train_valid_test_split(impute_data, target = "churn_probability", 
                                                                            train_size=0.7, valid_size=0.2, test_size=0.1, random_state=5)

In [6]:
# Hyperparameter Tuning for Random Forest Regressor
from sklearn.model_selection import RandomizedSearchCV

# Hyperparameters:
n_estimators = [int(i) for i in np.linspace(start = 100, stop = 500, num = 10)]

max_features =['auto','sqrt']

max_depth = [int(i) for i in np.linspace(10,50, num = 5)]
max_depth.append(None)

min_samples_split = [2,5,10]

param_grid = {"n_estimators": n_estimators, "max_features": max_features, "max_depth": max_depth, 
              "min_samples_split": min_samples_split}

In [ ]:
# Defining the model
RF2 = RandomForestClassifier()

# Hyperparameter tuning using rssandomizedsearch
modelRF2 = RandomizedSearchCV(estimator=RF2, param_distributions=param_grid, n_iter=10, cv=3, verbose=100, random_state=7, n_jobs=-1)

# Fitting the model
modelRF2.fit(X_train3, y_train3)

# Getting the best parameters
modelRF2.best_params_

In [93]:
# Training using the Random Forest Regressor
import timeit
# Creating the model
modelRF = RandomForestClassifier(n_estimators=500, max_depth=15, random_state=7, class_weight="balanced")

start = timeit.default_timer()

# Fitting the model
modelRF.fit(X_train3,y_train3)

stop = timeit.default_timer()

print('Time: ', stop - start)  

RandomForestClassifier(class_weight='balanced', max_depth=15, n_estimators=500,
                       random_state=7)

Time:  86.28100066600018


In [70]:
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score
# Predicting Values for the training set
predsRFTrain = modelRF.predict(X_train3)
print("Training Confusion Matrix:")
confusion_matrix(y_train3, predsRFTrain)
print("Training Accuracy:")
accuracy_score(y_train3, predsRFTrain)

# Predicting the values for validation set
predsRF = modelRF.predict(X_valid3)

# Validation error
print("\nValidation Confusion Matrix:")
confusion_matrix(y_valid3, predsRF)
# Validation error
print("\nValidation Accuracy:")
accuracy_score(y_valid3, predsRF)

# Predicting the values for the test set
predsRF2 = modelRF.predict(X_test3)

print("\nTest Confusion Matrix:")
confusion_matrix(y_test3, predsRF2)
print("\nTest Accuracy:")
accuracy_score(y_test3, predsRF2)

Training Confusion Matrix:


array([[42922,  1047],
       [  161,  4869]])

Training Accuracy:


0.9753464356415437


Validation Confusion Matrix:


array([[12224,   342],
       [  443,   991]])


Validation Accuracy:


0.9439285714285715


Test Confusion Matrix:


array([[6127,  205],
       [ 203,  465]])


Test Accuracy:


0.9417142857142857

In [90]:
# Splitting the data into Training, validation and test sets
X_train4, y_train4, X_valid4, y_valid4, X_test4, y_test4 = train_valid_test_split(df2, target = "churn_probability", 
                                                                            train_size=0.7, valid_size=0.2, test_size=0.1, random_state=5)

In [94]:
# Training using the Random Forest Regressor
import timeit
# Creating the model
modelRF = RandomForestClassifier(n_estimators=500, max_depth=15, random_state=7, class_weight="balanced")

start = timeit.default_timer()
# Fitting the model
modelRF.fit(X_train4,y_train4)

stop = timeit.default_timer()

print('Time: ', stop - start)  

RandomForestClassifier(class_weight='balanced', max_depth=15, n_estimators=500,
                       random_state=7)

Time:  49.89910429200063


In [99]:
from sklearn.metrics import classification_report
# Predicting Values for the training set
predsRFTrain = modelRF.predict(X_train4)
print("Training Confusion Matrix:")
print(confusion_matrix(y_train4, predsRFTrain))
print("Training Accuracy:")
print(classification_report(y_train4, predsRFTrain))

# Predicting the values for validation set
predsRF = modelRF.predict(X_valid4)

# Validation error
print("\nValidation Confusion Matrix:")
print(confusion_matrix(y_valid4, predsRF))
# Validation error
print("\nValidation Accuracy:")
print(classification_report(y_valid4, predsRF))

# Predicting the values for the test set
predsRF2 = modelRF.predict(X_test4)

print("\nTest Confusion Matrix:")
print(confusion_matrix(y_test4, predsRF2))
print("\nTest Accuracy:")
print(classification_report(y_test4, predsRF2))

Training Confusion Matrix:
[[42943  1026]
 [  245  4785]]
Training Accuracy:
              precision    recall  f1-score   support

         0.0       0.99      0.98      0.99     43969
         1.0       0.82      0.95      0.88      5030

    accuracy                           0.97     48999
   macro avg       0.91      0.96      0.93     48999
weighted avg       0.98      0.97      0.97     48999


Validation Confusion Matrix:
[[12207   359]
 [  440   994]]

Validation Accuracy:
              precision    recall  f1-score   support

         0.0       0.97      0.97      0.97     12566
         1.0       0.73      0.69      0.71      1434

    accuracy                           0.94     14000
   macro avg       0.85      0.83      0.84     14000
weighted avg       0.94      0.94      0.94     14000


Test Confusion Matrix:
[[6116  216]
 [ 199  469]]

Test Accuracy:
              precision    recall  f1-score   support

         0.0       0.97      0.97      0.97      6332
         1